src/tests/psyGen_test.py

```diff
@@ -810,3 +815,794 @@ def test_invalid_reprod_pad_size():
+    arg_f1_write_1 = schedule.children[0].children[0].arguments.args[1]
+    arg_f1_write_2 = schedule.children[1].children[0].arguments.args[1]
+    assert arg_f1_write_1._dependent_arg(arg_f1_write_2)
+
+
+def test_argument_find_argument():
+    '''Check that the find_argument method returns the first dependent
+    argument in a list of nodes, or None if none are found'''
+    _, invoke_info = parse(
+        os.path.join(BASE_PATH, "15.3.4_multi_axpy_invoke.f90"),
+        distributed_memory=True, api="dynamo0.3")
+    psy = PSyFactory("dynamo0.3", distributed_memory=True).create(invoke_info)
+    invoke = psy.invokes.invoke_list[0]
+    schedule = invoke.schedule
+    # 1: returns none if none found
+    f1_first_read = schedule.children[0].children[0].arguments.args[1]
+    # a) empty node list
+    assert not f1_first_read._find_argument([])
+    # b) check many reads
+    call_nodes = schedule.calls()
+    assert not f1_first_read._find_argument(call_nodes)
+    # 2: returns first dependent kernel arg when there are many
+    # dependencies (check first read returned)
+    f3_write = schedule.children[3].children[0].arguments.args[3]
+    f3_first_read = schedule.children[0].children[0].arguments.args[2]
+    result = f3_write._find_argument(call_nodes)
+    assert result == f3_first_read
+    # 3: haloexchange node
+    _, invoke_info = parse(
+        os.path.join(BASE_PATH, "15.1_builtin_and_normal_kernel_invoke.f90"),
+        distributed_memory=True, api="dynamo0.3")
+    psy = PSyFactory("dynamo0.3", distributed_memory=True).create(invoke_info)
+    invoke = psy.invokes.invoke_list[0]
+    schedule = invoke.schedule
+    # a) kern arg depends on halo arg
+    m2_read_arg = schedule.children[3].children[0].arguments.args[4]
+    m2_halo_field = schedule.children[2].field
+    result = m2_read_arg._find_argument(schedule.children)
+    assert result == m2_halo_field
+    # b) halo arg depends on kern arg
+    result = m2_halo_field._find_argument([schedule.children[3].children[0]])
+    assert result == m2_read_arg
+    # 4: globalsum node
+    _, invoke_info = parse(
+        os.path.join(BASE_PATH, "15.10.1_sum_field_builtin.f90"),
+        distributed_memory=True, api="dynamo0.3")
+    psy = PSyFactory("dynamo0.3", distributed_memory=True).create(invoke_info)
+    invoke = psy.invokes.invoke_list[0]
+    schedule = invoke.schedule
+    # a) globalsum arg depends on kern arg
+    kern_asum_arg = schedule.children[3].children[0].arguments.args[0]
+    glob_sum_arg = schedule.children[2].scalar
+    result = kern_asum_arg._find_argument(schedule.children)
+    assert result == glob_sum_arg
+    # b) kern arg depends on globalsum arg
+    result = glob_sum_arg._find_argument([schedule.children[3].children[0]])
+    assert result == kern_asum_arg
+
+
+@pytest.mark.xfail(reason="gh_readwrite not yet supported in PSyclone")
+def test_globalsum_arg():
+    '''Check that the globalsum argument is defined as gh_readwrite and
```

Comment:
gh_sum data is zero'ed first. The globalsum object takes partial sums from each MPI process, sums them together and updates them with the summed value on all MPI processes. It therefore has gh_readwrite access.